src/main/scala/is/hail/io/bgen/BgenBlockReader.scala

```diff
@@ -153,3 +77,59 @@ class BgenBlockReader(job: Configuration, split: FileSplit) extends IndexedBinar
     }
   }
 }
+
+class BgenBlockReaderV12(job: Configuration, split: FileSplit) extends BgenBlockReader[BgenRecordV12](job, split) {
+  override def createValue(): BgenRecordV12 = new BgenRecordV12(bState.compressed, bState.nSamples, tolerance)
+
+  override def next(key: LongWritable, value: BgenRecordV12): Boolean = {
+    if (pos >= end)
+      false
+    else {
+      val lid = bfis.readLengthAndString(2)
+      val rsid = bfis.readLengthAndString(2)
+      val chr = bfis.readLengthAndString(2)
+      val position = bfis.readInt()
+
+      val nAlleles = bfis.readShort()
+      assert(nAlleles >= 2, s"Number of alleles must be greater than or equal to 2. Found $nAlleles alleles for variant '$lid'")
+      val nAltAlleles = nAlleles - 1
+
+      val ref = bfis.readLengthAndString(4)
+      val altAlleles = Array.ofDim[String](nAltAlleles)
+
+      var altIndex = 0
+      while (altIndex < nAltAlleles) {
+        altAlleles(altIndex) = bfis.readLengthAndString(4)
+        altIndex += 1
+      }
+
+      val recodedChr = chr match {
+        case "23" => "X"
+        case "24" => "Y"
+        case "25" => "X"
+        case "26" => "MT"
+        case x => x
+      }
+
+      val variant = Variant(recodedChr, position, ref, altAlleles)
+
+      val dataSize = bfis.readInt()
+      val uncompressedSize = if (bState.compressed) bfis.readInt() else dataSize
+
+      val bytesInput =
+        if (bState.compressed)
+          bfis.readBytes(dataSize - 4)
```

Comment:
The compressed size includes uncompressedSize?  You might combine these two ifs:

```
  val (uncompressedSize, bytesInput) = ...
```